### PR TITLE
Jena update + Event Recorder

### PR DESCRIPTION
**Pull request by tnavatar**
I'm sorry for this massive PR. _Mea culpa, mea culpa, mea maxima culpa._

I'm doing a few things here that would have been better to split into multiple PRs, but at this point better to get it reviewed and integrated. These things, in no specific order:

* Update Jena to 4.5
  * The biggest aspect to this related to the PR below is that handling of JSON-LD changes significantly. This requires the following changes:
    * JSON-LD context for Gene Validity updated with prefix interpolation removed, although Kyle had a suggestion for making this work with the new JSON-LD 1.1 standard
    * JSON-LD functionality in ClinVar transformer removed; this has been already replaced in a different branch by Kyle's work.
    * References to old JSON-LD parsers removed from database.* ns
    * Lucene Suggest bumped to track Lucene version in Jena
* Adding Event Recorder
  * This takes input from streams and persists to a RocksDB database. The primary intention is to allow one to compare the effects of code changes against prior results from event processing.
  * Adding this required changes to the way Model objects are frozen; the prior approach failed for models over 60kB in size or so.
  * Refactoring to the RocksDB interface was also done to support key-scans on keys derived from byte arrays rather than Clojure data. The event recorder needs to use byte-array based keys to maintain ordering of events within a topic.
* Legacy topic data cleaned up

### CODE REVIEW

File: src/genegraph/transform/gene_validity_refactor.clj [hunk 1]
```diff
@@ -48,8 +48,7 @@
                construct-earliest-articles
                construct-secondary-contributions
                construct-variant-score
-               construct-unscoreable-evidence
-               )
+               construct-unscoreable-evidence)
```
Comment (toneillbroad) [inline src/genegraph/transform/gene_validity_refactor.clj:51]: Not a big deal, but I like the paren on it's own separate line here to do the nature of testing in the repl by commenting out all but a few of the constructs.

**Review by toneillbroad — APPROVED** [review]

File: src/genegraph/transform/clinvar/common.clj [hunk 2]
```diff
@@ -12,7 +12,7 @@
            (java.io StringWriter ByteArrayInputStream)
            (java.nio.charset Charset)
            (org.apache.jena.rdf.model Model)
-           (org.apache.jena.riot.writer JsonLDWriter)
+           ;; (org.apache.jena.riot.writer JsonLDWriter) ; can remove? =tristan
```
Comment (tbl3rd) [inline src/genegraph/transform/clinvar/common.clj:15]: Apparently yes?

File: src/genegraph/transform/gene_validity_refactor.clj (hunk 1 above)
Comment (tbl3rd) [reply src/genegraph/transform/gene_validity_refactor.clj:51]: Try using `#_`.

File: src/genegraph/transform/jsonld/common.clj [hunk 3]
```diff
@@ -76,29 +78,31 @@
...
+                                        ;(.set context JsonLDWriter/JSONLD_FRAME frame-str)
+       (log/trace :msg "Setting jsonld options")
+                                        ; TODO does nothing
+                                        ;(.setOmitGraph jsonld-options true)
+                                        ; TODO does nothing
+                                        ;(.setProcessingMode jsonld-options "JSON_LD_1_1")
+       (.set context JsonLDWriter/JSONLD_OPTIONS jsonld-options)
+       (log/trace :msg "Writing framed jsonld")
+       (.write writer sw ds prefix-map base-uri context)
+       (.toString sw)))))
```
Comment (tbl3rd) [inline src/genegraph/transform/jsonld/common.clj:108]: Again `#_` here would diff less than using `comment`.

File: src/genegraph/util/fs.clj [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+(defn ensure-target-directory-exists!
+  "Create directory at PATH if it does not already exist.
+  Return false if directory cannot be created or already
+  exists as something other than a directory."
+  [path]
+  (let [dir (io/file path)]
+    (if (and (.exists dir) (.isDirectory dir))
+      true
+      (.mkdir dir))))
```
Comment (tbl3rd) [inline src/genegraph/util/fs.clj:13]: Try `clojure.java.io/make-parents` instead.

File: test/genegraph/rocksdb_test.clj [hunk 5]
```diff
@@ -23,9 +23,18 @@
       (sut/rocks-put-multipart-key! db ["keyprefix1" "keysuffix1"] "value3")
       (sut/rocks-put-multipart-key! db ["keyprefix1" "keysuffix2"] "value4")
       (sut/rocks-put-multipart-key! db ["keyprefix2" "keysuffix2"] "value5")
+      (is (= 2 (-> (sut/prefix-seq db "keyprefix1") count)))
+      (is (= 1 (-> (sut/prefix-seq db "keyprefix2") count)))
```
Comment (tbl3rd) [inline test/genegraph/rocksdb_test.clj:27]: Why not just `(count (sut/prefix-seq db "keyprefix1"))`?
That way you see expected `count` right after the integer.

**Review by tbl3rd — COMMENTED** [review]
Some comments, but I still don't understand enough.

File: src/genegraph/transform/gene_validity_refactor.clj [hunk 6]
```diff
@@ -118,8 +117,8 @@
             ;; "Candidate gene sequencing" "gcixform:CandidateGeneSequencing"
             ;; "Exome/genome or all genes sequenced in linkage region" "gcixform:ExomeSequencing"
 
-            "Candidate gene sequencing" "SEPIO:0004543"
-            "Exome/genome or all genes sequenced in linkage region" "SEPIO:0004541"
+            "Candidate gene sequencing" "http://purl.obolibrary.org/obo/SEPIO_0004543"
```
Comment (theferrit32) [inline src/genegraph/transform/gene_validity_refactor.clj:120]: I believe these should be able to use the `SEPIO:` prefix if the definition of the `SEPIO` term in the context above explicitly sets `@prefix: true`.

```
"SEPIO": {"@id": "http://purl.obolibrary.org/obo/SEPIO_", "@prefix": true}
```

See this thread: https://github.com/w3c/json-ld-syntax/issues/329#issuecomment-592090824

**Review by theferrit32 — APPROVED** [review]

File: src/genegraph/sink/event_recorder.clj [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  Returns event, in case further processing on stream is needed."
+  [event]
+  (if (::record-event event)
+    (try 
+      (rocksdb/rocks-put-raw-key! 
+       event-database
+       (event-key event)
+       (-> event
+           (dissoc :io.pedestal.interceptor.chain/stack ; remove non-freezable keys
+                   :io.pedestal.interceptor.chain/execution-id)))
```
[the remark this answers is not in the document]
Comment (tnavatar) [reply src/genegraph/sink/event_recorder.clj:53]: In this case; Clojure fns,

File: src/genegraph/util/fs.clj (hunk 4 above)
Comment (tnavatar) [reply src/genegraph/util/fs.clj:13]: Good suggestion; I think that doesn't quite do the same thing at this, but one could phase this out by calling make-parents prior to opening the file in question, I think.

**Pull request merged**